app/models/app_remote_kill.rb

```diff
@@ -0,0 +1,77 @@
+class AppRemoteKill < ActiveRecord::Base
+  belongs_to :app
+  belongs_to :split
+
+  attribute :first_bad_version, :app_version
+  attribute :fixed_version, :app_version
+
+  validates :app, :split, :reason, :override_to, :first_bad_version, presence: true
+  validates :reason, uniqueness: { scope: %i(app split) }
+
+  validate :override_to_must_exist
+  validate :fixed_version_must_be_greater_than_first_bad_version
+  validate :must_not_overlap_existing
+
+  scope :affecting, ->(app_build, override: false, overridden_at: nil) do
+    where(
+      arel_table[:app_id].eq(app_build.app_id)
+      .and(arel_table[:first_bad_version].lteq(app_build.version))
+      .and(arel_fixed_version_is_null_or_greater_than(app_build.version))
+      .and(
+        Arel::Nodes::Grouping.new(
+          arel_is_false(override)
+          .or(arel_table[:updated_at].gt(overridden_at))
```

Comment:
The idea of the `override` and `overridden_at` args to affecting is that we'll consider a given remote kill not to affect us if override evaluates to true and overridden_at is more recent than the remote kill's last modification.

But the cool thing is that they can take ruby value objects like `false`, `nil`, or a `TimeWithZone`, or they can take arel expressions, meaning we can have other scopes call us with join columns accessible via the outer context of a subselect. So that's what we do here. We can test it simply here, but leverage it for sweet multi-table where conditions in calling scopes.

Reply:
@samandmoore the convo you had about remote kill being valuable for developer prerelease builds that want to make breaking API changes without revving the API version numbers is making me think that remote kills should not respect hard assignments at all regardless of recency. It would elevate remote kills to un-overrideable even by devs. Relative to the safety of not confusing an internal tester who might flip the flag on again and then see wildly broken app state, I’m now pretty bearish on allowing overrides to remote kills. If you want to see around a remote kill, you can cut yourself a local build with the flag hardcoded.

Aligns better with the seeming finality of a word like kill too (though kills can be edited/removed, they’d apply brutally when in place).

Reply:
Also the override behavior in local can be looser than on TT server once we have TT fake server to alleviate the potential pain of a dev trying to work on a bug fix for a feature that has an open-ended remote kill and potentially accidentally committing the gate for a not-yet-100%-weighted feature hard coded open.

Reply:
Given the way these PRs are stacked, I’m gonna suggest we review this code as it lies regardless and add the behavior change in a subsequent PR, and leave an architecture decision record about it (I introduced the repo’s first ADR in a subsequent PR too).